### PR TITLE
Add repo_id for attachment

### PR DESCRIPTION
**Pull request by lunny**
When create a new issue or comment and paste/upload an attachment/image, it will not assign an issue id before submit. So if user give up the creating, the attachments will lost key feature and become dirty content. We don't know if we need to delete the attachment even if the repository deleted.

This PR add a repo_id in attachment table so that even if a new upload attachment with no issue_id or release_id but should have repo_id. When deleting a repository, they could also be deleted.

### CODE REVIEW

File: routers/web/repo/attachment.go [hunk 1]
```diff
@@ -16,20 +16,21 @@ import (
...
 // UploadReleaseAttachment response for uploading release attachments
 func UploadReleaseAttachment(ctx *context.Context) {
-	uploadAttachment(ctx, setting.Repository.Release.AllowedTypes)
+	uploadAttachment(ctx, ctx.Repo.Repository.ID, setting.Repository.Release.AllowedTypes)
 }
 
 // UploadAttachment response for uploading attachments
-func uploadAttachment(ctx *context.Context, allowedTypes string) {
+func uploadAttachment(ctx *context.Context, repoID int64, allowedTypes string) {
 	if !setting.Attachment.Enabled {
```
Comment (6543) [inline routers/web/repo/attachment.go:34]: I think we should move this code :point_down: into **services/attachment/** too

File: models/repo.go [hunk 2]
```diff
@@ -530,12 +530,7 @@ func (repo *Repository) DeleteWiki() error {
 }
 
 func (repo *Repository) deleteWiki(e Engine) error {
```
Comment (6543) [inline models/repo.go]: this will make func repo.DeleteWiki() not work properly (L528)

Comment (6543) [reply models/repo.go]: I propose to delete this func here copy-paste sql code into `DeleteRepository` and the current function into `DeleteWiki`

File: models/migrations/v193.go [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+		RepoID     int64  `xorm:"INDEX"` // this should not be zero
+		IssueID    int64  `xorm:"INDEX"` // maybe zero when creating
+		ReleaseID  int64  `xorm:"INDEX"` // maybe zero when creating
+		UploaderID int64  `xorm:"INDEX DEFAULT 0"`
+	}
+	if err := x.Sync2(new(Attachment)); err != nil {
+		return err
+	}
+
+	if _, err := x.Exec("UPDATE attachment set repo_id = (SELECT repo_id FROM issue WHERE issue.id = issue_id) WHERE issue_id > 0"); err != nil {
```
Comment (6543) [inline models/migrations/v193.go]: this sql looks valid ;)

but a migration test is nice to ensure it will stay so :D

File: models/repo.go (hunk 2 above)
Comment (6543) [reply models/repo.go]: UPDATE: we shuld find func who use repo.DeleteWiki() and switch them to wiki_service.DeleteWiki(repo)

File: services/wiki/wiki.go [hunk 4]
```diff
@@ -366,3 +366,13 @@ func DeleteWikiPage(doer *models.User, repo *models.Repository, wikiName string)
 
 	return nil
 }
+
+// DeleteWiki removes the actual and local copy of repository wiki.
+func DeleteWiki(repo *models.Repository) error {
+	if err := repo.DeleteWiki(); err != nil {
```
Comment (6543) [inline services/wiki/wiki.go]: use `models.UpdateRepositoryUnits(repo, nil, []models.UnitType{UnitTypeWiki})` to be able to rm repo.DeleteWiki() func

File: models/repo.go (hunk 2 above)
Comment (lunny) [reply models/repo.go]: I have replaced all places.

Comment (6543) [reply models/repo.go]: @lunny commited my changes :)

File: models/migrations/v193_test.go [hunk 5]
```diff
@@ -0,0 +1,72 @@
...
+		UploaderID int64  `xorm:"INDEX DEFAULT 0"`
+	}
+
+	// Prepare and load the testing database
+	x, deferable := prepareTestEnv(t, 0, new(Attachment))
+	if x == nil || t.Failed() {
+		defer deferable()
+		return
+	}
+	defer deferable()
```
Comment (6543) [inline models/migrations/v193_test.go]:
```suggestion
	x, deferable := prepareTestEnv(t, 0, new(Attachment))
	defer deferable()
	if x == nil || t.Failed() {
		return
	}
```

File: models/migrations/v193.go (hunk 3 above)
Comment (6543) [inline models/migrations/v193.go]:
```suggestion
	if _, err := x.Exec("UPDATE `attachment` set repo_id = (SELECT repo_id FROM `issue` WHERE `issue`.id = `attachment`.issue_id) WHERE `attachment`.issue_id > 0"); err != nil {
```

File: models/migrations/v193.go [hunk 6]
```diff
@@ -0,0 +1,33 @@
...
+		return err
+	}
+
+	if _, err := x.Exec("UPDATE attachment set repo_id = (SELECT repo_id FROM release WHERE release.id = release_id) WHERE release_id > 0"); err != nil {
```
Comment (6543) [inline models/migrations/v193.go]:
```suggestion
	if _, err := x.Exec("UPDATE `attachment` set repo_id = (SELECT repo_id FROM `release` WHERE `release`.id = `attachment`.release_id) WHERE `attachment`.release_id > 0"); err != nil {
```

File: models/fixtures/attachment.yml [hunk 7]
```diff
@@ -81,6 +90,8 @@
 -
   id: 10
   uuid: a0eebc99-9c0b-4ef8-bb6d-6bb9bd380a20
+  repo_id: 1
+  release_id: 5
```
Comment (6543) [inline models/fixtures/attachment.yml]:
```suggestion
  repo_id: 0 # TestGetAttachment/NotLinked
```

**Review by 6543 — APPROVED** [review]

**Review by axifive — APPROVED** [review]

**Comment by 6543** [comment]
:rocket:

**Pull request merged**